noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -1,76 +1,120 @@
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute in the teardown function how much of AA the user needs
+    /// to pay to the `private_fee_recipient` and how much of it will be refunded back. Note that this is computed
+    /// based on an exchange rate between AA and fee juice.
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
+    ///
+    /// With this scheme a user has privately paid for the tx fee with an arbitrary AA (e.g. could be a stablecoin).
+    ///
+    /// ***Note:***
+    /// This flow allows us to pay for the tx with msg_sender's private balance of AA and hence msg_sender's identity
+    /// is not revealed. We do, however, reveal the `max_fee`.
     #[private]
-    fn fee_entrypoint_private(amount: Field, asset: AztecAddress, nonce: Field) {
+    fn fee_entrypoint_private(max_fee: Field, nonce: Field) {
         // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
-        let settings = storage.settings.read();
+        let config = storage.config.read();
 
-        assert(asset == settings.other_asset);
-
-        Token::at(asset).setup_refund(settings.admin, context.msg_sender(), amount, nonce).call(
-            &mut context,
-        );
+        Token::at(config.accepted_asset)
+            .setup_refund(config.private_fee_recipient, context.msg_sender(), max_fee, nonce)
+            .call(&mut context);
         context.set_as_fee_payer();
     }
 
+    /// Pays for the tx fee with msg_sender's public balance of accepted asset (AA). The maximum fee a user is willing
+    /// to pay is defined by `max_fee` and is denominated in AA.
+    ///
+    /// ## Overview
+    /// The refund flow works as follows:
+    /// 1. We pull the `max_fee` from the user's balance of the AA to this contract (requires setting an authwit),
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute how much of AA the user needs to pay to
+    /// the `private_fee_recipient` and how much of it will be refunded back. Note that this is computed based on
+    /// an exchange rate between AA and fee juice.
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
+    ///
+    /// ***Note:***
+    /// AA funds sent by the users to this contract stay in this contract and are not transferred
+    /// to the `private_fee_recipient`. In the private flow we needed a separate `private_fee_recipient` as this
+    /// contract does not have keys associated with it. In production we would want to have a method allowing for
+    /// pulling of these funds from this contract.
     #[private]
-    fn fee_entrypoint_public(amount: Field, asset: AztecAddress, nonce: Field) {
-        FPC::at(context.this_address())
```

Comment:
Nuked the prepare_fee function as it was superfluous --> we could just enqueue the call to the token directly from the private function (which saves 1 kernel iteration).